### ISSUE TITLE
Implement WebContentsDelegate::RunBluetoothChooser

### ISSUE THREAD

**Issue by barchard** [electron/electron#5565]
* Electron version:
We are using node 5.10.0, Chrome 49.0.2623.75, and Electron 1.0.1. 

* Operating system:
Mac OS X 10.11.3 (15D21)

Trying to use the [Web Bluetooth APIs](https://developers.google.com/web/updates/2015/07/interact-with-ble-devices-on-the-web?hl=en#scan-for-bluetooth-devices) but when I run the Web Bluetooth demo code, the callbacks fail with the error:

`Argh! NotFoundError: User cancelled the requestDevice() chooser.`

main.js code:
```
// Module to control application life.
const {app} = electron;

// @link https://github.com/electron/electron/blob/master/docs/api/chrome-command-line-switches.md
app.commandLine.appendSwitch('enable-web-bluetooth', true);
```

Index.html code:

```
<a href="#" onclick="run();" class="scanDevices">Scan</a>
    <script>
    function run() {
		var filters = [{
    services: [0x1800, 0x1801, 0x1812]
  }];
		  console.log('Requesting Bluetooth Device...');
  navigator.bluetooth.requestDevice({filters: filters})
  .then(device => {
    console.log('> Name:             ' + device.name);
    console.log('> Id:               ' + device.id);
    console.log('> UUIDs:            ' + device.uuids.join('\n' + ' '.repeat(20)));
    console.log('> Connected:        ' + device.gatt.connected);
  })
  .catch(error => {
    console.log('Argh! ' + error);
  });
  }
    </script>
```

Reference:
https://github.com/electron/electron/blob/master/atom/browser/atom_permission_manager.cc#L93

**Comment by deepak1556** [comment]
Does the above example work with chrome ? I tried the example on linux and was able to reproduce the error only when a service was not supported by the device, other than that behavior was consistent with chromium.

Although on linux there are some initializations required to enable bluetooth, will fix it in a PR.

**Comment by beaufortfrancois** [comment]
FYI @scheib @g-ortuno @jyasskin

**Comment by barchard** [comment]
On Chrome, I get a different error:

```
Argh! SecurityError: requestDevice() called with a filter containing a blacklisted UUID. https://goo.gl/4NeimX
```

**Comment by barchard** [comment]
Changing the code to:

```
<html>
	<head>
	</head>
	<body>
	<a href="#" onclick="run();" class="scanDevices">Scan</a>
    <script>
    function run() {
        var filters = [{
    services: [0x1800]
  }];
          console.log('Requesting Bluetooth Device...');
  navigator.bluetooth.requestDevice({filters: filters})
  .then(device => {
    console.log('> Name:             ' + device.name);
    console.log('> Id:               ' + device.id);
    console.log('> UUIDs:            ' + device.uuids.join('\n' + ' '.repeat(20)));
    console.log('> Connected:        ' + device.gatt.connected);
  })
  .catch(error => {
    console.log('Argh! ' + error);
  });
  }
    </script>
	</body>
</html>
```

Results in the same (original) error:
```
Requesting Bluetooth Device...
chrome.html:20 Argh! NotFoundError: User cancelled the requestDevice() chooser.
```

**Comment by g-ortuno** [comment]
I took a look at the Electron code. There doesn't seem to be an implementation of [`WebContentsDelegate::RunBluetoothChooser`](https://code.google.com/p/chromium/codesearch#chromium/src/content/public/browser/web_contents_delegate.h&sq=package:chromium&type=cs&l=356&rcl=1464074483), so we would fallback to the first seen device or, if we can't find any devices, fail with the error you are currently seeing.

**Comment by barchard** [comment]
Hi

Thanks. Would this be considered a bug or feature request in Electron? Chrome? Libchromium? My code?

Thanks

**Comment by g-ortuno** [comment]
Feature request in Electron. Chrome has its own implementation [here](https://code.google.com/p/chromium/codesearch#chromium/src/chrome/browser/ui/browser.cc&sq=package:chromium&type=cs&l=1325&rcl=1464074483).

**Comment by zcbenz** [comment]
@g-ortuno When `WebContentsDelegate::RunBluetoothChooser` is not implemented Chromium will choose the first device automatically. We should implement it in Electron, but it won't solve op's problem though since WebBluetooth is not working for him even in Chrome browser.

**Comment by barchard** [comment]
happy to try and test. I'm running os x at the moment but target deployment is Win 10. I can test in Win10 later this week or next.

**Comment by jyasskin** [comment]
@zcbenz No devices advertise 1800 or 1801, so filtering for them will always fail to find devices, and [1812 is indeed blacklisted](https://github.com/WebBluetoothCG/registries/blob/master/gatt_blacklist.txt#L9). On Chrome 50 and above on OSX, @barchard should have seen an empty chooser dialog which they'd have to cancel to get the NotFoundError. They can file bugs at https://crbug.com/ if something listed at https://github.com/WebBluetoothCG/web-bluetooth/blob/gh-pages/implementation-status.md as done isn't working.

**Comment by barchard** [comment]
@jyasskin definitely never seen an empty chooser dialog.

**Comment by jyasskin** [comment]
Yep, happens for me too, with your code. Follow along at https://crbug.com/614568.

**Comment by deepak1556** [comment]
@jyasskin on linux i tried @barchard code with `0x1800` and `0x1801` and i was able to retrieve the uuid, i actually filtered service only after verifiying with `bluetoothctl`.

```
[6178:0525/082026:INFO:CONSOLE(25)] "Requesting Bluetooth Device...", source: http://localhost:8000/render-bg.html (25)
[6178:0525/082026:WARNING:bluetooth_dispatcher_host.cc(1247)] No Bluetooth chooser implementation; falling back to first device.
[6178:0525/082026:INFO:CONSOLE(28)] "> Name:             OnePlus One", source: http://localhost:8000/render-bg.html (28)
[6178:0525/082026:INFO:CONSOLE(29)] "> Id:               MDxKvCVV3r4pHg2LHJOKIg==", source: http://localhost:8000/render-bg.html (29)
[6178:0525/082026:INFO:CONSOLE(30)] "> UUIDs:            00001800-0000-1000-8000-00805f9b34fb
                    00001801-0000-1000-8000-00805f9b34fb", source: http://localhost:8000/render-bg.html (30)
[6178:0525/082026:INFO:CONSOLE(31)] "> Connected:        false", source: http://localhost:8000/render-bg.html (31)
```

Also another issue is when i connect to the device and try to read a characteristic the promise wont resolve, this happens with chromium too and thereafter any other navigation crashes.

```
[6415:0525/082647:FATAL:render_frame_host_impl.cc(350)] Check failed: this == frame_tree_node_->current_frame_host(). 
#0 0x7f7c5a53553e base::debug::StackTrace::StackTrace()
#1 0x7f7c5a54efdb logging::LogMessage::~LogMessage()
#2 0x7f7c5c5eeb66 content::RenderFrameHostImpl::GetLastCommittedOrigin()
#3 0x7f7c5c528e5a content::BluetoothDispatcherHost::OnGATTServerDisconnect()
#4 0x7f7c5c527eee content::BluetoothDispatcherHost::OnMessageReceived()
#5 0x7f7c5a5364ec base::debug::TaskAnnotator::RunTask()
#6 0x7f7c5a554c8b base::MessageLoop::RunTask()
#7 0x7f7c5a554f98 base::MessageLoop::DeferOrRunPendingTask()
#8 0x7f7c5a5551bb base::MessageLoop::DoWork()
#9 0x7f7c5a526c89 base::MessagePumpGlib::Run()
#10 0x7f7c5a571570 base::RunLoop::Run()
#11 0x7f7c5c53c0d9 content::BrowserMainLoop::MainMessageLoopRun()
#12 0x7f7c5c53bed4 content::BrowserMainLoop::RunMainMessageLoopParts()
#13 0x7f7c5c53ecad <unknown>
#14 0x7f7c5c53770c content::BrowserMain()
#15 0x7f7c5c4b51f2 <unknown>
#16 0x7f7c5c4b3f40 content::ContentMain()
#17 0x00000084bf72 main
#18 0x7f7c5199e741 __libc_start_main
```

The crash doesnt happen with chromium.

**Comment by deepak1556** [comment]
Update: the crash doesnt happen anymore with chrome 51 upgrade in electron.

**Comment by beaufortfrancois** [comment]
@barchard As seen in https://bugs.chromium.org/p/chromium/issues/detail?id=614568#c1, you might want to replace `<a>` with `<button>`.

```html
<html>
    <head>
    </head>
    <body>
    <button onclick="run();" class="scanDevices">Scan</button>
    <script>
    function run() {
        var filters = [{
    services: [0x1800]
  }];
          console.log('Requesting Bluetooth Device...');
  navigator.bluetooth.requestDevice({filters: filters})
  .then(device => {
    console.log('> Name:             ' + device.name);
    console.log('> Id:               ' + device.id);
    console.log('> UUIDs:            ' + device.uuids.join('\n' + ' '.repeat(20)));
    console.log('> Connected:        ' + device.gatt.connected);
  })
  .catch(error => {
    console.log('Argh! ' + error);
  });
  }
    </script>
    </body>
</html>

**Comment by barchard** [comment]
@beaufortfrancois @zcbenz I can confirm, changing from an `<a>` to a `<button>` in Chrome on OS X. Shall I now test Electron? If so, what build is recommended?

**Comment by beaufortfrancois** [comment]
The code sample you've shown should be working in Chrome Canary for Mac OS X: https://www.google.com/chrome/browser/canary.html?platform=mac
